### ISSUE TITLE
Doesn't work properly on kernel 3.13

### ISSUE THREAD

**Issue by Linerd** [Vzaa/bsnoop#1]
I have compiled the code (I used the old commit 'compiles on 3.13'), and used insmod the insert the snoop.ko, but it seems that the module is not seeing the traffic. 0 connections served. Anything else I should do to test the code?

**Comment by Vzaa** [comment]
I worked on this a while ago to test some idea but that idea ended up being unnecessarily complicated to implement. So it's been a while since I looked at the code but I guess I can try to help you.

What is your setup? If your intent is to run it on layer 3 gateways as the original project that won't work with that specific commit because apparently before the initial import I actually changed the netfilter hook to NF_BR_FORWARD as my intention was to get this running on a layer 2 bridge. So it won't work on the original setup on that commit even though it'll compile. Maybe you can try to change these values back to the original values from http://sourceforge.net/projects/bsnoop/:
https://github.com/Vzaa/bsnoop/blob/bfed985b8920e5dbbfc134230cd2d6e67941527e/snoop.c#L1038

I was able to get it to run on layer 2 at some point but I'm not sure it was fully working. On that specific commit the part where it's looking for the IP headers are all wrong.

**Comment by Linerd** [comment]
Alright, I shall try and see if it works. Thanks for the quick reply!

**Comment by Linerd** [comment]
Well I tried to change the net filter hook to NF_INET_FORWARD and the .pf to PF_INET, this time I guess the code is working, coz when a flow passed by, the router machine crashed…

**Comment by Vzaa** [comment]
Unfortunately I currently don't have a setup to test it :/ I haven't tested this at layer 3 at all but like I said was able to get it running on layer 2 like I mentioned.

**Comment by Linerd** [comment]
What is your setup for layer2? Which commit were you testing?

**Comment by Vzaa** [comment]
I tested this on a desktop PC with Ubuntu 14.04. I had a bridge interface that was bridging two Ethernet interfaces. One connected to the internet gateway and the other to a wireless access point. I was testing out different ideas and committing the changes somewhat randomly. So I can't properly point out to a single 'good' commit. Maybe I can prepare and test a "clean" version next week when i have the time and let you know. That is if you're still interested in working on this.

**Comment by Linerd** [comment]
Yeah, I’d interested. Actually I have got the code running on kernel 3.13 last night, but somehow the performance is way worse than that on kernel 2.6. I didn’t actually change much, but the cache miss number is ridiculously high. I set the cache size to be 1000, which literally has no cache miss on kernel 2.6, but gets more cache misses than retransmitted. I am guessing there is something wrong with the hashing? Do you have any idea?
